### PR TITLE
swarm-derive/: Add where clause of behaviour to generated out event

### PR DESCRIPTION
**Pull request by mxinden**
# Description

When generating the `OutEvent` for a `NetworkBehaviour`, add the `where` clause of the
`NetworkBehaviour` `struct` to the generated `enum`.

## Links to any relevant issues

Follow up to https://github.com/libp2p/rust-libp2p/pull/2792

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Review by thomaseizinger — APPROVED** [review]

**Comment by elenaf9** [comment]
Won't this cause an error in case of
```rust
    #[derive(NetworkBehaviour)]
    struct Foo<T>
    where
        T: Copy
    {
        ping: libp2p::ping::Ping,
        #[behaviour(ignore)]
        bar: T,
    }
```
which will generate 

```rust
pub enum FooEvent<T> // Error: unused Parameter
where
    T: Copy
{
    Ping(<libp2p::ping::Ping as NetworkBehaviour>::OutEvent)
}
```
Am I missing something?

**Comment by mxinden** [comment]
@elenaf9 that is correct. Good catch. Though why would someone use the `OutEvent` mechanism in combination with `#[behaviour(ignore)]`?

Once we remove `NetworkBehaviourEventProcess` (https://github.com/libp2p/rust-libp2p/pull/2751) I don't think there is any use for `poll_method` nor `ignore` at all. See https://github.com/libp2p/rust-libp2p/pull/2751#discussion_r930451630.

With that in mind, I would leave this as is. Do you agree @elenaf9?

**Comment by elenaf9** [comment]
> Though why would someone use the OutEvent mechanism in combination with #[behaviour(ignore)]?
> 
> Once we remove NetworkBehaviourEventProcess (https://github.com/libp2p/rust-libp2p/pull/2751) I don't think there is any use for poll_method nor ignore at all. See https://github.com/libp2p/rust-libp2p/pull/2751#discussion_r930451630.

I am a bit confused. Why would we not need `#[behaviour(ignore)]` anymore? Isn't it still needed to signal the derive macro that this field is not a network behaviour? E.g. when implementing the methods of the `NetworkBehaviour` trait and we pass the event to our inner behaviours?
And with `OutEvent` I would think we stilll need it to mark that this field won't generate an event:
- If the `OutEvent` is provided by the user, we should not need the trait bound `From<<FIELD as NetworkBehaviour>::OutEvent>` for this field
- If the `OutEvent` is generated, we should not generated an enum variant for this field. See above code example.

@mxinden  what am I missing here?

**Comment by mxinden** [comment]
> I am a bit confused. Why would we not need `#[behaviour(ignore)]` anymore?

Let me rephrase: 

With `NetworkBehaviourEventProcess` users sometimes needed an `events: VecDeque` which they could push events into from `NetworkBehaviourEventProcess` and pop events from within their custom `poll` method. `events` would not implement `NetworkBehaviour` and thus needed the `#[behaviour(ignore)]`.

With the removal of `NetworkBehaviourEventProcess` there is no more need for fields on the root `NetworkBehaviour` that do not themselves implement `NetworkBehaviour`, at least I can not think of a use-case. 

Can you think of a use case for `#[behaviour(ignore)]` and a custom `poll` method once `NetworkBehaviourEventProcess` is removed @elenaf9?

**Comment by elenaf9** [comment]
Thanks for elaborating @mxinden!  Makes sense now. 
> Can you think of a use case for #[behaviour(ignore)] and a custom poll method once NetworkBehaviourEventProcess is removed @elenaf9?

I guess the only use case is if the wrapping `NetworkBehaviour` wants to implement some additional logic on top of the inner behaviours and / or hide some of the implementation details about them. But this can also be implemented on top of the `Swarm`, so I agree that we don't need `#[behaviour(ignore)]` or `#[behaviour(poll =  ..)]` anymore. 

> With that in mind, I would leave this as is. Do you agree @elenaf9?

Yep :)
